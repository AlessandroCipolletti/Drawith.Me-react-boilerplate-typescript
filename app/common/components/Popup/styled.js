import styled from 'styled-components'
import { parseColorHexToRgbaString } from 'common/utils/'

import { SoftShadow } from 'common/styled/Shadows'

export const Overlay = styled.div`
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  z-index: 600;
  opacity: ${props => props['data-visible'] ? 1 : 0};
  transition-duration: ${props => props.theme.timing.fadeAnimation.string};
  background-color: ${props => props.theme.UIColors.overlayBackgroundColor};
  display: flex;
  align-items: center;
  justify-content: center;
`

export const Wrapper = styled(SoftShadow)`
  max-width: 90vw;
  max-height: 90vh;
  min-width: ${props => props.theme.spacing.size11};
  min-height: ${props => props.theme.spacing.size11};
  overflow: auto;
  border-radius: ${props => props.theme.spacing.size1};
  z-index: 650;
  background-color: ${props => props.theme.palette.gray[9]};
  & > *:first-child {
    max-width: 90vw;
    max-height: 90vh;
  }
`

Wrapper.CloseButton = styled.div`
  position: absolute;
  right: ${props => `-${props.theme.spacing.size3}`};
  top: ${props => `-${props.theme.spacing.size3}`};
  background-color: ${props => props.theme.palette.white};
  border-radius: 50%;
  z-index: 100;
  cursor: pointer;
  &:hover {
    background: ${props => parseColorHexToRgbaString(props.theme.palette.gray[1], 0.5)};
  }
`

export const Header = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: ${props => props.theme.spacing.size8};
  display: flex;
  align-items: center;
  box-sizing: border-box;
  padding: 0px ${props => props.theme.spacing.size5} ;
  background: ${props => props.theme.gradients.primaryColorLight};
`

Header.Title = styled.span`
  color: ${props => props.theme.palette.gray[9]};
  font-size: ${props => props.theme.fontSizing.size5};
  user-select: none;
  font-weight: 600;
`

export const Content = styled.div`
  position: relative;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  padding-top: ${props => props.theme.spacing.size8};
`
