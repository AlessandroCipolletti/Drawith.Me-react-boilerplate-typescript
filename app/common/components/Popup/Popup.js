import React, { memo } from 'react'
import PropTypes from 'prop-types'
import Shake from 'common/styled/Shake'

import Theme from 'common/Theme'

import Icon from 'common/components/Icon'
import { Overlay, Wrapper, Header, Content } from './styled'

const Popup = ({
  callback, children, showClose = false, allowClose = true,
}) => {
  const [visible, setVisible] = React.useState(false)
  const [isShaking, setIsShaking] = React.useState(false)
  const overlayEl = React.useRef(null)
  const closeButtonEl = React.useRef(null)

  const close = React.useCallback(() => {
    setVisible(false)
    setTimeout(callback, Theme.timing.fadeAnimation.ms)
  }, [callback])

  const shake = React.useCallback(() => {
    setIsShaking(true)
    setTimeout(() => { setIsShaking(false) }, 1000)
  }, [])

  const onOverlayClick = React.useCallback((e) => {
    if (e.target === overlayEl.current && visible) {
      if (allowClose) {
        close()
      } else {
        shake()
      }
    }
  }, [visible, allowClose])

  React.useEffect(() => {
    setVisible(true)
  }, [])

  return (
    <Overlay
      ref={overlayEl}
      onClick={onOverlayClick}
      data-testid="overlay"
      data-visible={visible}
    >
      <Shake isShaking={isShaking}>
        <Wrapper>
          {children}
          {showClose && allowClose &&
            <Wrapper.CloseButton ref={closeButtonEl}>
              <Icon name='exit' size={Theme.spacing.size5} color={Theme.palette.gray[9]} onClick={close} />
            </Wrapper.CloseButton>
          }
        </Wrapper>
      </Shake>
    </Overlay>
  )
}

Popup.propTypes = {
  callback: PropTypes.func.isRequired,
  children: PropTypes.any,
  showClose: PropTypes.bool,
  allowClose: PropTypes.bool,
}

Popup.Header = Header
Popup.Content = Content

export default memo(Popup)
