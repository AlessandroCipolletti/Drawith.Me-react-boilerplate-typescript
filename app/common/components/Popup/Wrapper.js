import styled from 'styled-components'

const Wrapper = styled.div`
  position: absolute;
  top: 48%;
  left: 50%;
  max-width: 90%;
  max-height: 90%;
  min-width: ${props => props.theme.spacing.size12};
  min-height: ${props => props.theme.spacing.size12};
  overflow: auto;
  transform: translate3d(-50%, -50%, 0px);
  padding: ${props => props.theme.spacing.size5};
  border-radius: ${props => props.theme.spacing.size2};
  z-index: 1050;
  background-color: ${props => props.theme.palette.gray[9]};
`

Wrapper.defaultProps = {
  theme: {
    palette: {
      gray: {
        9: 'white',
      },
    },
    spacing: {
      size2: '8px',
      size5: '24px',
      size12: '256px',
    },
  },
}

export default Wrapper
