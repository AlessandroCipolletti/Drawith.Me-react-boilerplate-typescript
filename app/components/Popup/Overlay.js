import styled from 'styled-components'

const Overlay = styled.div`
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  z-index: 1000;
  background-color: ${props => props.theme.UIColors.popupBackgroundColor};
`

Overlay.defaultProps = {
  theme: {
    UIColors: {
      popupBackgroundColor: 'transparent',
    },
  },
}

export default Overlay
