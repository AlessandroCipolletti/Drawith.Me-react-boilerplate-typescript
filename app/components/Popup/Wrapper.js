import styled from 'styled-components'

const Wrapper = styled.div`
  position: absolute;
  top: 48%;
  left: 50%;
  max-width: 90%;
  max-height: 90%;
  min-width: 100px;
  min-height: 100px;
  overflow: auto;
  transform: translate3d(-50%, -50%, 0px);
  padding: ${props => props.theme.spacing.s};
  border-radius: ${props => props.theme.spacing.xxs};
  z-index: 1050;
  background-color: ${props => props.theme.palette.whiteColor};
`

Wrapper.defaultProps = {
  theme: {
    palette: {
      whiteColor: 'white',
    },
    spacing: {
      xxs: '8px',
      s: '24px',
    },
  },
}

export default Wrapper
