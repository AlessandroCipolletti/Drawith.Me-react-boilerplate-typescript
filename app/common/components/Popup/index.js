import React, { memo } from 'react'
import PropTypes from 'prop-types'

import { FadeIn, FadeOut } from 'animate-css-styled-components'
import Theme from 'common/theme'

import Overlay from './Overlay'
import Wrapper from './Wrapper'

function Popup({
  callback, children,
}) {
  const [visible, setVisible] = React.useState(true)

  const close = () => {
    if (visible) {
      setVisible(false)
      setTimeout(callback, Theme.timing.fadeAnimation.ms)
    }
  }

  const content = (
    <Overlay onClick={close} data-testid="overlay">
      <Wrapper>
        {children}
      </Wrapper>
    </Overlay>
  )

  if (visible) {
    return (
      <FadeIn duration={Theme.timing.fadeAnimation.string}>
        {content}
      </FadeIn>
    )
  }

  return (
    <FadeOut duration={Theme.timing.fadeAnimation.string}>
      {content}
    </FadeOut>
  )
}

Popup.propTypes = {
  callback: PropTypes.func.isRequired,
  children: PropTypes.any.isRequired,
}

export default memo(Popup)
