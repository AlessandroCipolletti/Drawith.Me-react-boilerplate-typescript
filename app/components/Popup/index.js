import React, { memo, useState } from 'react'
import PropTypes from 'prop-types'

import { FadeIn, FadeOut } from 'animate-css-styled-components'
import Theme from 'common/theme'

import Overlay from './Overlay'
import Wrapper from './Wrapper'

function Popup({
  callback,
}) {
  // TODO Test Unitaires
  const [visible, setVisible] = useState(true)

  const close = () => {
    if (visible) {
      setVisible(false)
      setTimeout(callback, parseFloat(Theme.timing.fadeAnimation) * 1000)
    }
  }

  const content = (
    <Overlay onClick={close}>
      <Wrapper>
        CONTENT
      </Wrapper>
    </Overlay>
  )

  if (visible) {
    return (
      <FadeIn duration={Theme.timing.fadeAnimation}>
        {content}
      </FadeIn>
    )
  }

  return (
    <FadeOut duration={Theme.timing.fadeAnimation}>
      {content}
    </FadeOut>
  )
}

Popup.propTypes = {
  callback: PropTypes.func.isRequired,
}

export default memo(Popup)
