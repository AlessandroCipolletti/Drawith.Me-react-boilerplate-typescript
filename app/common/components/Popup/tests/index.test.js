/**
 *
 * Tests for Popup
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react'
import { render, fireEvent, getByTestId } from 'react-testing-library'

// Import de Theme Provider
import { ThemeProvider } from 'styled-components'
import Theme from 'common/Theme'

import Popup from '../index'

describe('<Popup />', () => {
  const setState = jest.fn()
  const setStateSpy = jest.spyOn(React, 'useState')

  const renderComponent = (props = {}) =>
    render(
      <ThemeProvider theme={Theme}>
        <Popup {...props}>
          Content
        </Popup>
      </ThemeProvider>,
    )

  beforeEach(() => {

  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')
    const callback = () => {}

    renderComponent({ callback })

    expect(spy).not.toHaveBeenCalled()
  })

  it('Expect log an error in console if called without callback prop', () => {
    const spy = jest.spyOn(global.console, 'error')

    renderComponent()

    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('Should render and match the snapshot', () => {
    const callback = () => {}

    const {
      container: { firstChild },
    } = renderComponent({ callback })

    expect(firstChild).toMatchSnapshot()
  })

  it('Should prevent multiple calls of close callback', async () => {
    const callback = jest.fn()

    const { container } = renderComponent({ callback })
    const overlay = getByTestId(container, 'overlay')
    fireEvent.click(overlay)
    fireEvent.click(overlay)
    fireEvent.click(overlay)
    await new Promise((resolve) => setTimeout(resolve, Theme.timing.fadeAnimation.ms * 1.5))

    expect(callback).toHaveBeenCalledTimes(1)
  })

  it('Should set `visible` state to false on Overlay click via setVisible(false)', () => {
    const callback = () => {}
    setStateSpy.mockImplementation((init) => [init, setState])

    const { container } = renderComponent({ callback })
    const overlay = getByTestId(container, 'overlay')
    fireEvent.click(overlay)

    expect(setState).toHaveBeenCalledTimes(1)
    expect(setState).toHaveBeenCalledWith(false)
  })

  it('Should re-render et match the shapshot for hiding after calling setVisible(false)', () => {
    const callback = () => {}

    const { asFragment, container } = renderComponent({ callback })

    const overlay = getByTestId(container, 'overlay')
    fireEvent.click(overlay)

    expect(asFragment()).toMatchSnapshot()
  })

})
