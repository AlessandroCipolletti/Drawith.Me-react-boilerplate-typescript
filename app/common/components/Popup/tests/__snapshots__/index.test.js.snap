// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> Should re-render et match the shapshot for hiding after calling setVisible(false) 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-kGXeez gvkiBY"
    direction="normal"
    display="block"
  >
    <div
      class="Overlay-qtskko-0 fTghja"
      data-testid="overlay"
    />
    <div
      class="Wrapper-sc-1d51d2j-0 bIVduR"
    >
      Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Popup /> Should render and match the snapshot 1`] = `
<div
  class="sc-bdVaJa sc-kGXeez gvkiBY"
  direction="normal"
  display="block"
>
  <div
    class="Overlay-qtskko-0 fTghja"
    data-testid="overlay"
  />
  <div
    class="Wrapper-sc-1d51d2j-0 bIVduR"
  >
    Content
  </div>
</div>
`;
